flask-app/app.py

```diff
@@ -47,19 +54,103 @@ def proxy_request_to_brouter(parameters):
+
+
+def get_region(model, region_id):
+    db_session = DBSession()
+    region = db_session.query(model).filter_by(id_=region_id).first()
+    if not region:
+        return f"{model.__name__} with given id not found", 400
+    return region.to_dict()
+
+
+@app.route("/api/v1/voivodeships")
+def get_all_voivodeships():
+    return get_regions(Voivodeship)
+
+
 @app.route("/api/v1/counties")
 def get_all_counties():
-    db_session = DBSession()
-    return {"counties": [county.to_dict() for county in db_session.query(County).all()]}
+    return get_regions(County)
+
+
+@app.route("/api/v1/voivodeships/<string:voivodeship_id>")
+def get_voivodeship(voivodeship_id):
+    return get_region(Voivodeship, voivodeship_id)
 
 
 @app.route("/api/v1/counties/<string:county_id>")
 def get_county(county_id):
+    return get_region(County, county_id)
+
+
+def get_county_ids_from_cases():
     db_session = DBSession()
-    county = db_session.query(County).filter_by(id_=county_id).first()
-    if not county:
-        return "county with given id not found", 400
-    return county.to_dict()
+    columns = db_session.query(CasesRecord.county_id).distinct()
+    return [county_id for county_id in columns]
+
+
+@app.route("/api/v1/cases/")
+def get_cases():
+    '''Get most recent cases for each county.'''
+    ids = get_county_ids_from_cases()
+    return {"cases": [get_cases_for(county_id) for county_id in ids]}
+
+
+@app.route("/api/v1/cases/<string:date_str>")
+def get_cases_from(date_str):
+    '''Get most recent cases up till given date for each county.'''
+    ids = get_county_ids_from_cases()
+
+    records_dicts = []
+    for county_id in ids:
+        try:
+            cases = get_cases_from_for(date_str, county_id)
+        except NoResultFound:
+            continue
+        records_dicts.append(cases)
+
+    if not records_dicts:
+        raise NoResultFound
+
+    return {"cases": records_dicts}
+
+
+@app.route("/api/v1/cases/for/<string:county_id>")
+def get_cases_for(county_id):
+    '''Get most recent cases entry for county of given id.'''
+    db_session = DBSession()
+    record = db_session.query(CasesRecord) \
+        .filter_by(county_id=county_id) \
+        .order_by(CasesRecord.updated.desc()).first()
+
+    if not record:
+        raise NoResultFound
+
+    return record.to_dict()
+
+
+@app.route("/api/v1/cases/<string:date_str>/for/<string:county_id>")
+def get_cases_from_for(date_str, county_id):
+    '''Get most recent cases entry up till given date for county of given id.'''
+    db_session = DBSession()
+    try:
+        dt = datetime.datetime.strptime(date_str, r"%Y-%m-%d").date()
+    except ValueError:
+        return "Wrong date provided. Please use the proper format: 'DD-MM-YYYY'", 400
+
+    record = db_session.query(CasesRecord) \
+        .filter_by(county_id=county_id) \
+        .order_by(CasesRecord.updated.desc()) \
+        .filter(CasesRecord.updated <= dt).first()
```

Comment:
We should discuss if this should be `<= dt` or `== dt`

Reply:
As we discussed in PMs, I changed this to `== dt`.
In case we decide that oldest entries are also needed, we could just change `c.updated == date` to `c.updated <= date` (d76e8b1 - line 113).